### PR TITLE
Add deployment of Azure Bakery (Kubernetes based)

### PR DESCRIPTION
**Pull request by ciaranevans**
## What I am changing

TL;DR: This PR adds the capability to deploy a complete Pangeo Forge Azure Bakery (k8s based) from a clean slate, all you need is an Azure Account (with what I believe to be pretty wide stretching permissions 😅 ). This PR creates all the Terraform resources & scripts needed to deploy the base infrastructure to support AKS and then the application of the Prefect K8s configuration on top. A Pangeo Forge Recipe that is setup as a flow is also provided to register & run against the Bakery to exercise it. (See [How you can test it](#how-you-can-test-it) for an explanation of the state of play when actually running this recipe)

(Sorry it's big 😬 )

## How I did it

* Terraform changes in `terraform/`:
  * `main.tf`:
    * Bumped terraform provider version to `0.15.1`
    * Using `azurerm` as the remote state backend
    * Add configuration to create the storage infrastructure required, the AKS cluster, and the log workspace
  * `outputs.tf`:
    * Export various values for use further in deployment (namely resource names)
* Scripts in `scripts/`:
  * `replace_or_insert_value.py`:
    * Allows us to modify `.env` whilst we perform deployment operations, rather than outputting values for the Bakery operator to have to copy and paste and inevitably forget
    * For a given environment variable name and value, if the variable **is** present in `.env`, we update its value, otherwise we insert a line with `variable_name=value`
  * `setup_remote_state.sh`:
    * This is required as a one-off setup to create the necessary storage account and resource group to allow us to use Azure as the remote state for Terraform, it's the ol' chicken and egg situation
  * `build_and_push_image.sh`:
    * Uses the deployed ACR details to build the Prefect Agent & Worker image stored in `images/Dockerfile` and pushes it to ACR
  * `retrieve_flow_storage_values.sh`:
    * Sets the values of `FLOW_STORAGE_CONTAINER` and `FLOW_STORAGE_CONNECTION_STRING` within `.env` via Terraform outputs and the Azure CLI
* Docker Image for Prefect Agent & Workers in `images/`:
  * `Dockerfile`:
    * Builds an image based on `prefecthq/prefect:0.14.19-python3.8` and installs the requirements in `requirements.txt`
  * `requirements.txt`:
    * What it says on the tin, all the dependencies required for the Bakery Images
* Prefect Agent K8s configuration in `prefect_agent_conf.yaml`
  * Generated from the Prefect CLI and modified to refer to environment variables we provide
  * Notably I have set `DELETE_FINISHED_JOBS=False` as otherwise a successful Prefect Job is lost from the UI in Azure
* Example Recipe -> Flow for testing in `flow_test/oisst_recipe.py`:
  * This was cobbled together with @sharkinsspatial's help, it **almost** runs properly (See [How you can test it](#how-you-can-test-it) for an explanation of the state of play when actually running this recipe)
* Several `Makefile` changes:
  * Add the following steps:
    * `install` - installs the `poetry` dependencies
    * `setup-remote-state` - Calls `setup_remote_state.sh`
    * `destroy-remote-state` - Calls AZ CLI to destroy the remote state infra
    * `plan` - Runs a terraform plan
    * `apply` - Runs a terraform apply
    * `destroy` - Runs a terraform destroy
    * `configure-kubectl` - Points kubectl at the newly deployed AKS cluster
    * `setup-agent` - Uses kubectl to apply the Prefect Agent config to AKS
    * `retrieve-flow-storage-values` - Calls `retrieve_flow_storage_values.sh`
    * `build-and-push-image` - Calls `build_and_push_image.sh`
    * `deploy-bakery` - Wraps all calls needed to deploy the Bakery - Can be run over and over again.
* README improvements:
  * Add instructions for installing dependencies (and also updating the requirements required to develop on the project)
  * Modify instructions for setting up Azure Credentials to **actually** use the Service Principle we create
  * Remove notion of `deployment.tfvars.json` and replace with `.env` - explaining all the variables expected
  * Further breakdown of the `Makefile` contents
  * Add detailed breakdown of how to deploy the bakery
  * Add details on how to use Azure Logging
* Add `example.env` to show all the expected Environment Variables
* General changes:
  * Use `poetry` for python dependencies (Used to use Pyenv - See the AWS Bakery, but the dependency resolving just degrades so quickly)
  * New entries to `.gitignore`
  * Remove `init` step in CI linting job
  
## How you can test it

First, ensure you've run through all the steps in Development > Requirements & Getting Started and Deployment > Prerequisites.

You can then deploy the bakery with:

```bash
$ make deploy-bakery
```

Once it's deployed, you can register the Flow with:

```bash
$ poetry run dotenv run python3 flow_test/oisst_recipe.py
```

Then navigate to `cloud.prefect.io` and run the flow.

This is where we have issues, the flow runs several tasks and then fails with the following error:
```
Unexpected error: JSONDecodeError("Expecting ',' delimiter: line 20 column 1 (char 314)")
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/prefect/engine/runner.py", line 48, in inner
    new_state = method(self, state, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/prefect/engine/task_runner.py", line 865, in get_task_run_state
    value = prefect.utilities.executors.run_task_with_timeout(
  File "/usr/local/lib/python3.8/site-packages/prefect/utilities/executors.py", line 323, in run_task_with_timeout
    return task.run(*args, **kwargs)  # type: ignore
  File "flow_test/oisst_recipe.py", line 26, in wrapper
  File "/usr/local/lib/python3.8/site-packages/rechunker/executors/prefect.py", line 39, in run
    return self.stage.func()
  File "/Users/ciaran/Library/Caches/pypoetry/virtualenvs/pangeo-forge-azure-bakery-IMqFot_V-py3.8/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 307, in prepare_target
  File "/usr/local/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 483, in expand_target_dim
    zgroup[dim][:] = 0
  File "/usr/local/lib/python3.8/site-packages/zarr/hierarchy.py", line 341, in __getitem__
    return Array(self._store, read_only=self._read_only, path=path,
  File "/usr/local/lib/python3.8/site-packages/zarr/core.py", line 158, in __init__
    self._load_metadata()
  File "/usr/local/lib/python3.8/site-packages/zarr/core.py", line 175, in _load_metadata
    self._load_metadata_nosync()
  File "/usr/local/lib/python3.8/site-packages/zarr/core.py", line 190, in _load_metadata_nosync
    meta = decode_array_metadata(meta_bytes)
  File "/usr/local/lib/python3.8/site-packages/zarr/meta.py", line 33, in decode_array_metadata
    meta = parse_metadata(s)
  File "/usr/local/lib/python3.8/site-packages/zarr/meta.py", line 27, in parse_metadata
    meta = json_loads(s)
  File "/usr/local/lib/python3.8/site-packages/zarr/util.py", line 35, in json_loads
    return json.loads(ensure_text(s, 'ascii'))
  File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.8/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 20 column 1 (char 314)
```

We are currently investigating this, we believe `adlfs` is to blame.

### CODE REVIEW

**Review by TomAugspurger — COMMENTED** [review]
Thanks. I just gave it a quick glance. Approving so that you can merge when your ready.

Anything in particular I should give a close review?

**Comment by ciaranevans** [comment]
As discussed in the call on Friday, @TomAugspurger is happy with this to go in, whilst we investigate the `adlfs` issues.
